### ISSUE TITLE
NMS time limit in val.py

### ISSUE THREAD

**Issue by 733ly** [ultralytics/yolov5#13702]
### Search before asking

- [x] I checked the docs, issues, and discussions and could not find an answer.

### Question

When I run val.py with the trained weights, I encounter the NMS time limit. When I try to reduce the batch size, the NMS warning decreases a bit, but the detection results seem to improve in terms of metrics. However, if I want to conduct a comparative experiment, shouldn't the batch size be kept the same? But I also suspect that the appearance of the NMS warning has led to a decline in performance. So I don't know how to make a choice.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @733ly, thank you for your interest in YOLOv5 🚀! Please visit our ⭐️ [Tutorials](https://docs.ultralytics.com/yolov5/) to get started, where you can find quickstart guides for simple tasks like [Custom Data Training](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data/) all the way to advanced concepts like [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution/).  
This is an automated response 🤖—an Ultralytics engineer will also assist soon.

If this is a 🐛 Bug Report, please provide a **minimum reproducible example (MRE)** to help us debug it (exact command used to run `val.py`, your args, and anything needed to reproduce the NMS time limit message).  
If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) ✅.

## Requirements
[**Python>=3.8.0**](https://www.python.org/) with all [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) installed including [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). To get started:
```bash
git clone https://github.com/ultralytics/yolov5  # clone
cd yolov5
pip install -r requirements.txt  # install
```

## Environments
YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://bit.ly/yolov5-paperspace-notebook"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

## Status
<a href="https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml"><img src="https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml/badge.svg" alt="YOLOv5 CI"></a>  
If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are currently passing. CI tests verify correct operation of YOLOv5 [training](https://github.com/ultralytics/yolov5/blob/master/train.py), [validation](https://github.com/ultralytics/yolov5/blob/master/val.py), [inference](https://github.com/ultralytics/yolov5/blob/master/detect.py), [export](https://github.com/ultralytics/yolov5/blob/master/export.py) and [benchmarks](https://github.com/ultralytics/yolov5/blob/master/benchmarks.py) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

To help us triage the “NMS time limit” behavior quickly 🧩, please add:
-Your exact `val.py` command (all args) and any modifications to [val.py](https://github.com/ultralytics/yolov5/blob/master/val.py)
-The full console log showing the warning
-Hardware details (GPU/CPU), OS, Python + PyTorch versions
-A note on whether the issue reproduces on the same weights with the same settings across runs

**Comment by pderrenger** [comment]
Batch size shouldn’t change your mAP; if your metrics move when you change `--batch`, it’s likely because the “NMS time limit” warning means NMS is bailing out early and you’re effectively evaluating with missing/partial detections, so for a fair comparison keep the same eval settings and tweak validation to avoid the warning (e.g., slightly raise `--conf-thres` and/or reduce the number of candidates via `--imgsz`/`--max-det` if you’re using it), then use whatever batch size fits your hardware. See the [YOLOv5 validation section](https://github.com/ultralytics/yolov5#validation) for the available `val.py` args and the [Non-Maximum Suppression (NMS) glossary](https://www.ultralytics.com/glossary/non-maximum-suppression-nms) for background on what’s happening.

**Comment by 733ly** [comment]
> Batch size shouldn’t change your mAP; if your metrics move when you change `--batch`, it’s likely because the “NMS time limit” warning means NMS is bailing out early and you’re effectively evaluating with missing/partial detections, so for a fair comparison keep the same eval settings and tweak validation to avoid the warning (e.g., slightly raise `--conf-thres` and/or reduce the number of candidates via `--imgsz`/`--max-det` if you’re using it), then use whatever batch size fits your hardware. See the [YOLOv5 validation section](https://github.com/ultralytics/yolov5#validation) for the available `val.py` args and the [Non-Maximum Suppression (NMS) glossary](https://www.ultralytics.com/glossary/non-maximum-suppression-nms) for background on what’s happening.

So you mean that I should adjust parameters like max_det and iou, so that nms is reduced, and ensure that the rest of the model also operates under this environment, right?

**Comment by pderrenger** [comment]
Yes—those are post-processing/eval settings, so choose `conf/iou/max_det` (or, if needed, increase the NMS `time_limit` in the [YOLOv5 `non_max_suppression` implementation](https://github.com/ultralytics/yolov5/blob/master/utils/general.py)) to eliminate the timeout, then keep that exact eval config for every model you compare; you don’t need to retrain, and `--batch` can be whatever runs without hitting the warning.
